src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs

```diff
@@ -5,130 +5,188 @@
-                    resourceData.Failed++;
+                    httpData.Failed++;
                 }
 
                 if (pdEvent.Bytes > 0)
                 {
-                    resourceData.TotalBytes += pdEvent.Bytes;
-                    if (pdEvent.Bytes > resourceData.MaxBytes)
-                    {
-                        resourceData.MaxBytes = pdEvent.Bytes;
-                    }
+                    httpData.TotalBytes += pdEvent.Bytes;
                 }
 
                 if (pdEvent.HttpStatusCode.HasValue)
                 {
-                    if (!resourceData.StatusCodes.TryGetValue(pdEvent.HttpStatusCode.Value, out var count))
+                    if (!httpData.StatusCodes.TryGetValue(pdEvent.HttpStatusCode.Value, out var count))
                     {
                         count = 0;
                     }
-                    resourceData.StatusCodes[pdEvent.HttpStatusCode.Value] = count + 1;
+                    httpData.StatusCodes[pdEvent.HttpStatusCode.Value] = count + 1;
                 }
             }
         }
 
-        private static void ApplyTiming(ResourceTimingData timingData, TimeSpan duration)
+        private void ProtocolDiagnostics_NupkgCopiedEvent(ProtocolDiagnosticNupkgCopiedEvent ncEvent)
         {
-            timingData.Requests++;
-            timingData.TotalDuration += duration;
+            AddNupkgCopiedData(ncEvent, _data, _knownSources);
+        }
+
+        internal static void AddNupkgCopiedData(ProtocolDiagnosticNupkgCopiedEvent ncEvent, ConcurrentDictionary<string, Data> allData, HashSet<string> sources)
+        {
+            if (!sources.Contains(ncEvent.Source))
+            {
+                return;
+            }
+
+            var data = allData.GetOrAdd(ncEvent.Source, _ => new Data());
+
+            lock (data._lock)
+            {
+                data.NupkgCount++;
+                data.NupkgSize += ncEvent.FileSize;
+            }
         }
 
         public void Dispose()
         {
-            ProtocolDiagnostics.Event -= ProtocolDiagnostics_Event;
+            ProtocolDiagnostics.HttpEvent -= ProtocolDiagnostics_HttpEvent;
+            ProtocolDiagnostics.ResourceEvent -= ProtocolDiagnostics_ResourceEvent;
+            ProtocolDiagnostics.NupkgCopiedEvent -= ProtocolDiagnostics_NupkgCopiedEvent;
         }
 
-        public void SendTelemetry()
+        public async Task SendTelemetryAsync()
         {
             var parentId = _parentId.ToString();
             foreach (var kvp in _data)
             {
                 Data data = kvp.Value;
                 string source = kvp.Key;
-                if (!_sources.TryGetValue(kvp.Key, out PackageSource packageSource))
+                if (!_sources.TryGetValue(kvp.Key, out SourceRepository sourceRepository))
                 {
-                    packageSource = null;
+                    sourceRepository = new SourceRepository(new PackageSource(source), Repository.Provider.GetCoreV3());
```

Comment:
And update `_knownSources` ? Otherwise, seems subsequent Telemetry for the same source won't take advantage of the Dictionary.